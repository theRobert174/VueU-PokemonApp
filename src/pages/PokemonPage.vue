<template>
  <h1 v-if="!pokemon">Espere por favor ...</h1>
  <div v-else>
    <div style="display: relative; height: 110px;">
      <h1>Who's that Pokemon?</h1>
      <h1 class="back-title">Who's that Pokemon?</h1>
    </div>
    <PokemonPicture style="" :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>

    <template v-if="showAnswer" >
      <h2 class="fade-in" :class="{'glow-correct' : isCorrect, 'glow-incorrect' : !isCorrect}">{{message}}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'
// console.log(getPokemonOptions())

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
      return {
        pokemonArr: [],
        pokemon: null,
        showPokemon: false,
        message: '',
        showAnswer: false,
        isCorrect: false
      }
    },
    methods:{
      async mixPokemonArray(){
        this.pokemonArr = await getPokemonOptions()

        const rndInt = Math.floor(Math.random() * 4)
        this.pokemon = this.pokemonArr[rndInt]
      },

      checkAnswer(selectedId){
        // console.log('Respuesta de evento options', selectedId)
          this.showPokemon = true
          this.showAnswer = true

        if(selectedId === this.pokemon.id){
          this.isCorrect = true
          this.message = `Correcto, ${this.pokemon.name}`
        }
        else {
          this.isCorrect = false
          this.message = `Oops, era ${this.pokemon.name}`
        }
        //console.log('HolaMundo',this.showPokemon)
      },

      async newGame(){
        this.showPokemon = this.showAnswer = false
        this.pokemonArr = []
        this.pokemon = null
        await this.mixPokemonArray()
      }
    },
    mounted(){
      this.mixPokemonArray()
    }
}
</script>

<style scoped>
  @import url('http://fonts.cdnfonts.com/css/pokemon-hollow');
  h1{
    font-family: 'Pokemon Solid', sans-serif;
    color: gold;
    position: absolute;
    right: 37%;
    user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    -webkit-user-drag: none;
    -webkit-user-select: none;
  }
  .back-title{
    font-family: 'Pokemon Hollow', sans-serif;
    color: blue;
  }

  .glow-correct {
  text-align: center;
  -webkit-animation: glow-correct 1s ease-in-out infinite alternate;
  -moz-animation: glow-correct 1s ease-in-out infinite alternate;
  animation: glow-correct 1s ease-in-out infinite alternate;
  }

  @keyframes glow-correct {
    from {
      text-shadow: 0 0 10px #fff, 0 0 20px #fff, 0 0 30px #2ee600, 0 0 40px #2ee600, 0 0 50px #2ee600, 0 0 60px #2ee600, 0 0 70px #2ee600;
    }
    to {
      text-shadow: 0 0 20px #fff, 0 0 30px #7cff4d, 0 0 40px #7cff4d, 0 0 50px #7cff4d, 0 0 60px #7cff4d, 0 0 70px #7cff4d, 0 0 80px #7cff4d;
    }
  }

  .glow-incorrect {
  text-align: center;
  -webkit-animation: glow-incorrect 1s ease-in-out infinite alternate;
  -moz-animation: glow-incorrect 1s ease-in-out infinite alternate;
  animation: glow-incorrect 1s ease-in-out infinite alternate;
  }

  @keyframes glow-incorrect {
    from {
      text-shadow: 0 0 10px #fff, 0 0 20px #fff, 0 0 30px #e60000, 0 0 40px #e60000, 0 0 50px #e60000, 0 0 60px #e60000, 0 0 70px #e60000;
    }
    to {
      text-shadow: 0 0 20px #fff, 0 0 30px #ff4d4d, 0 0 40px #ff4d4d, 0 0 50px #ff4d4d, 0 0 60px #ff4d4d, 0 0 70px #ff4d4d, 0 0 80px #ff4d4d;
    }
  }
</style>