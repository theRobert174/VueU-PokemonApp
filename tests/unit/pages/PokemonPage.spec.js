import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from '@/pages/PokemonPage'
import {pokemons} from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {

    let wrapper
    let mixPokemonArraySpy

    beforeEach(() => {
        mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        wrapper = shallowMount(PokemonPage)
    })

    test('debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe llamar mixPokemonArray al montar', () => {
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  message: '',
                  showAnswer: false,
                  isCorrect: false
                }
              }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  message: '',
                  showAnswer: false,
                  isCorrect: false
                }
              }
        })


        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        //PokemonPicture debe de existir
        expect(picture.exists()).toBeTruthy()
        //PokemonOptions debe de existir
        expect(options.exists()).toBeTruthy()
        //PokemonPicture attribute pokemonid === 5'
        expect(picture.attributes('pokemonid')).toBe('5')
        //PokemonOptions attribute pokemons toBe true
        expect(options.attributes('pokemons')).toBeTruthy()
    })

    test('pruebas con checkAnswer', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  message: '',
                  showAnswer: false,
                  isCorrect: false
                }
              }
        })

        await wrapper.vm.checkAnswer(5)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(10)
        expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`)
    })
})