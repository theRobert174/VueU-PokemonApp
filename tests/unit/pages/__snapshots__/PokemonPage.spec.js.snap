// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `<h1>Espere por favor ...</h1>`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <div style="display: relative; height: 110px;">
    <h1>Who's that Pokemon?</h1>
    <h1 class="back-title">Who's that Pokemon?</h1>
  </div>
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
